### PR TITLE
added option for hdf5 History to read_result_from_file

### LINKED ISSUE #259: Fill result from history
When a run crashes, there should be a function to fill a result from history (if such a one has been saved to file via csv/hdf5, that is). Partly exists already in the pypesto.optimize.optimizer.full_result_from_history function, but the rest of the workflow, both from csv and hdf5, should be added, to create the result object and finally store everything in the result hdf5 file.

### PR DESCRIPTION
**Pull request by PaulJonasJost**
closes #259.

### CODE REVIEW

File: pypesto/objective/history.py [hunk 1]
```diff
@@ -1115,10 +1118,10 @@ def _compute_vals_from_trace(self):
         # etc
         max_init_iter = 3
         for it in range(min(len(self.history), max_init_iter)):
-            candidate = self.history.get_fval_trace(it)
+            candidate = self.history.get_fval_trace()[it]
```
Comment (yannikschaelte) [inline pypesto/objective/history.py:1121]: why is this necessary? passing `it` into the function was designed to allow for efficient access, e.g. not calculating the full trajectory, but only that specific index/indices. Not sure if csv/hdf5 use it, but could potentially.

Comment (PaulJonasJost) [reply pypesto/objective/history.py:1121]: csv uses it, but hdf5 did not that's why I changed it for now into this to be able to use the function in both file formats.

Comment (yannikschaelte) [reply pypesto/objective/history.py:1121]: oh, I see. That's a bug in the History class: It seems like when `it` was introduced, it was not properly propagated down to hdf5, only csv, therefore the function calls there raise a "Signature does not match base class" warning in pycharm. Could you modify the functions in the Hdf5History?

Comment (PaulJonasJost) [reply pypesto/objective/history.py]: updated

**Review by yannikschaelte — APPROVED** [review]

File: pypesto/optimize/optimizer.py [hunk 2]
```diff
@@ -205,17 +206,25 @@ def fill_result_from_objective_history(
...
+        history = Hdf5History.load(
+            id=identifier,
+            file=history_options.storage_file,
+        )
     else:
         raise NotImplementedError()
 
     opt_hist = OptimizerHistory(
-        history, history.get_x_trace(0),
+        history, history.get_x_trace()[0],
```
Comment (PaulJonasJost) [inline pypesto/optimize/optimizer.py:227]:
```suggestion
        history, history.get_x_trace(0),
```

File: pypesto/objective/history.py [hunk 3]
```diff
@@ -955,18 +960,18 @@ def _generate_hdf5_group(self, f: h5py.File = None):
         except OSError:
             pass
 
-    def _get_hdf5_entries(self, entry_id: str) -> Sequence:
+    def _get_hdf5_entries(self, entry_id: str,
+                          ix: Union[int, Sequence[int], None]) -> Sequence:
         """
         returns the entries for the key entry_id.
```
Comment (jvanhoefer) [inline pypesto/objective/history.py:966]: doc string for `Parameters`

**Review by jvanhoefer — APPROVED** [review]

File: test/base/test_history.py [hunk 4]
```diff
@@ -141,18 +143,25 @@ def check_reconstruct_history(self, start: OptimizerResult, id: str):
...
+                                    f'get_{entry}_trace')()
+            for iteration in range(len(original_trace)):
+                # comparing nan and None difficult
+                if original_trace[iteration] is None or np.isnan(
+                        original_trace[iteration]).all():
+                    continue
+                np.testing.assert_array_almost_equal(
+                    reconst_trace[iteration],
+                    original_trace[iteration],
+                    decimal=15
```
Comment (PaulJonasJost) [inline test/base/test_history.py:163]:
```suggestion
                    decimal=10
```

File: pypesto/objective/history.py (hunk 3 above)
Comment (PaulJonasJost) [reply pypesto/objective/history.py:966]:
```suggestion
        """
        Parameters
        ------------
        entry_id:
            The key whose trace is returned.
        ix:
            Index or list of indices of the iterations that will produce
            the trace.
            
        Returns
        --------
        The entries ix for the key entry_id.
```

File: pypesto/objective/history.py [hunk 5]
```diff
@@ -955,18 +960,28 @@ def _generate_hdf5_group(self, f: h5py.File = None):
...
         """
-        returns the entries for the key entry_id.
+        Parameters
+        ------------
+        entry_id:
+            The key whose trace is returned.
+        ix:
+            Index or list of indices of the iterations that will produce
+            the trace.
+
```
Comment (PaulJonasJost) [inline pypesto/objective/history.py:973]:
```suggestion

```

File: test/base/test_history.py [hunk 6]
```diff
@@ -141,18 +147,25 @@ def check_reconstruct_history(self, start: OptimizerResult, id: str):
...
+        for entry in history_entries:
+            original_trace = getattr(start.history,
+                                     f'get_{entry}_trace')()
+            reconst_trace = getattr(reconst_history,
+                                    f'get_{entry}_trace')()
+            for iteration in range(len(original_trace)):
+                # comparing nan and None difficult
+                if original_trace[iteration] is None or np.isnan(
+                        original_trace[iteration]).all():
+                    continue
```
Comment (yannikschaelte) [inline test/base/test_history.py:163]: better check that the same then also applies to the reconst_ one?

File: pypesto/objective/history.py [hunk 7]
```diff
@@ -955,18 +961,27 @@ def _generate_hdf5_group(self, f: h5py.File = None):
         except OSError:
             pass
 
-    def _get_hdf5_entries(self, entry_id: str) -> Sequence:
+    def _get_hdf5_entries(self, entry_id: str,
+                          ix: Union[int, Sequence[int], None]) -> Sequence:
```
Comment (yannikschaelte) [inline pypesto/objective/history.py:965]:
```suggestion
    def _get_hdf5_entries(
        self,
        entry_id: str,
        ix: Union[int, Sequence[int], None] = None,
    ) -> Sequence:
```

**Review by yannikschaelte — APPROVED** [review]
Looks good to me, thanks! @FFroehlich as this affects some of your code, you may want to have a look, but I think we can merge already.

File: test/base/test_history.py [hunk 8]
```diff
@@ -49,22 +49,28 @@ def check_history(self):
...
-                history_options=self.history_options
-            )
-
-            for istart, start in enumerate(result.optimize_result.list):
-                self.check_reconstruct_history(start, str(istart))
-                self.check_load_from_file(start, str(istart))
-                self.check_history_consistency(start)
+        for storage_type in ['.csv',
+                             '.hdf5',
+                             None]:
```
Comment (yannikschaelte) [inline test/base/test_history.py:54]:
```suggestion
        for storage_type in ['.csv', '.hdf5', None]:
```

File: test/base/test_history.py (hunk 6 above)
Comment (PaulJonasJost) [reply test/base/test_history.py:163]:
```suggestion
                    assert(original_trace[iteration] is None or 
                        np.isnanoriginal_trace[iteration]).all()
```

Comment (PaulJonasJost) [reply test/base/test_history.py:163]: like this @yannikschaelte ?

Comment (yannikschaelte) [reply test/base/test_history.py:163]: maybe separate None and nan, then it should be like the previous check
